src/update_handler.c

```diff
@@ -1921,6 +1947,162 @@ static gboolean img_to_boot_emmc_handler(RaucImage *image, RaucSlot *dest_slot,
 }
 #endif
 
+static gboolean check_if_area_is_clear(const gchar *device, guint64 start, gsize size, gboolean *clear, GError **error)
+{
+	gboolean res = FALSE;
+	g_autofree guchar *read_buf = NULL;
+	gint read_size = 512;
+	gint fd;
+
+	g_return_val_if_fail(device, FALSE);
+	g_return_val_if_fail(clear, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	read_buf = g_malloc0(read_size);
+
+	fd = g_open(device, O_RDONLY);
+	if (fd == -1) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Opening device failed: %s",
+				g_strerror(errno));
+		goto out;
+	}
+
+	if (lseek(fd, start, SEEK_SET) != (off_t) start) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Failed to set file to position %"G_GUINT64_FORMAT ": %s",
+				start, g_strerror(errno));
+		goto out;
+	}
+
+	*clear = TRUE;
+
+	while (size && *clear) {
+		gint read_count = 0;
+
+		if (size < (gsize) read_size)
+			read_size = size;
+
+		read_count = read(fd, read_buf, read_size);
+		if (read_count < 0)
+			goto out;
+
+		for (gint i = 0; i < read_count; i++)
+		{
+			if ((read_buf[i] != 0x00) && (read_buf[i] != 0xFF)) {
+				*clear = FALSE;
+				break;
+			}
+		}
+
+		size -= read_count;
+	}
+
+	res = TRUE;
+
+out:
+	if (fd >= 0)
+		g_close(fd, NULL);
+
+	return res;
+}
+
+static gboolean img_to_boot_raw_fallback_handler(RaucImage *image, RaucSlot *dest_slot, const gchar *hook_name, GError **error)
+{
+	gboolean res = FALSE;
+	GError *ierror = NULL;
+	guint64 region_size;
+	gboolean primary_clear;
+	struct part_desc {
+		const char *name;
+		struct boot_switch_partition partition;
+	} part_desc[2];
+	int first_part_desc_index = 0;
+	g_autoptr(GHashTable) vars = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, g_free);
+	const gsize header_size = 512;
+
+	g_return_val_if_fail(image, FALSE);
+	g_return_val_if_fail(dest_slot, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	region_size = dest_slot->region_size / 2;
```

Comment:
The name `region_size` is used for two different things here. Perhaps we should use `half_size` here.

Reply:
Good idea.